mackerel-plugin.go

```diff
@@ -0,0 +1,76 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+func main() {
+	os.Exit(run(os.Args))
+}
+
+const (
+	exitOK = iota
+	exitError
+)
+
+var helpReg = regexp.MustCompile(`--?h(?:elp)?`)
+
+//go:generate sh -c "perl tool/gen_mackerel_plugin.pl > mackerel-plugin_gen.go"
+func run(args []string) int {
+	var plug string
+	f, err := exec.LookPath(args[0])
+	if err != nil {
+		log.Println(err)
+		return exitError
+	}
+	fi, err := os.Lstat(f)
+	if err != nil {
+		log.Println(err)
+		return exitError
+	}
+	base := filepath.Base(f)
+	if fi.Mode()&os.ModeSymlink == os.ModeSymlink && strings.HasPrefix(base, "mackerel-plugin-") {
+		// if mackerel-plugin is symbolic linked from mackerel-plugin-memcached, run the memcached plugin
+		plug = strings.TrimPrefix(base, "mackerel-plugin-")
+	} else {
+		if len(args) < 2 {
+			printHelp()
+			return exitError
+		}
+		plug = args[1]
+		if helpReg.MatchString(plug) {
+			printHelp()
+			return exitOK
+		}
+		osargs := []string{f}
+		osargs = append(osargs, args[2:]...)
+		os.Args = osargs
+	}
+
+	err = runPlugin(plug)
+
+	if err != nil {
+		return exitError
+	}
+	return exitOK
+}
+
+var version, gitcommit string
+
+func printHelp() {
+	fmt.Printf(`mackerel-plugin %s
+
+Usage: mackerel-plugin <plugin> [<args>]
+
+Following plugins are available:
+    %s
+
+See `+"`mackerel-plugin <plugin> -h` "+`for more information on a specific plugin
```

Comment:
[nit]
filename of this binary (maybe `mackerel-plugin`) will be shown in each plugin's usage on called `mackerel-plugin <plugin> -h`.
Do you have any idea to solve this? (I think it's difficult…)
```
[github.com/mackerelio/mackerel-agent-plugins/build]$ ./mackerel-agent-plugins uptime -help
Usage of ./mackerel-agent-plugins:
  -metric-key-prefix string
    	Metric key prefix (default "uptime")
  -tempfile string
    	Temp file name
[github.com/mackerelio/mackerel-agent-plugins/build]$ ls -la ./mackerel-plugin-uptime
lrwxr-xr-x  1 astj  staff  22  3  6 16:12 ./mackerel-plugin-uptime@ -> mackerel-agent-plugins
[github.com/mackerelio/mackerel-agent-plugins/build]$ ./mackerel-plugin-uptime -help
Usage of ./mackerel-plugin-uptime:
  -metric-key-prefix string
    	Metric key prefix (default "uptime")
  -tempfile string
    	Temp file name
```